sdk/nodejs/provider/server.ts

```diff
@@ -350,8 +316,67 @@ class Server implements grpc.UntypedServiceImplementation {
             }
             resp.setState(structproto.Struct.fromJavaScript(state));
 
-            // Wait for RPC operations to complete and disconnect.
-            await runtime.disconnect();
+            // Wait for RPC operations to complete.
+            await runtime.waitForRPCs();
+
+            callback(undefined, resp);
+        } catch (e) {
+            console.error(`${e}: ${e.stack}`);
+            callback(e, undefined);
+        }
+    }
+
+    public async call(call: any, callback: any): Promise<void> {
+        // Serialize invocations of `construct` and `call` so that each call runs one after another, avoiding concurrent
+        // runs. We do this because `construct` and `call` modify global state to reset the SDK's runtime options.
+        // This is a short-term workaround to provide correctness, but likely isn't sustainable long-term due to the
+        // limits it places on parallelism. We will likely want to investigate if it's possible to run each invocation
+        // in its own context, possibly using Node's `createContext` API to avoid modifying global state:
+        // https://nodejs.org/api/vm.html#vm_vm_createcontext_contextobject_options
+        const res = this.constructCallQueue.then(() => this.callImpl(call, callback));
+        // tslint:disable:no-empty
+        this.constructCallQueue = res.catch(() => {});
+        return res;
+    }
+
+    async callImpl(call: any, callback: any): Promise<void> {
+        try {
```

Comment:
We need to wrap this in a global uncaught handler that resolves the gRPC call as was done in https://github.com/pulumi/pulumi/pull/7365

Reply:
I included this when I rebased against master.